### PR TITLE
Automagic pricing

### PR DESCRIPTION
**Pull request by vkuznecovas**
Pricing is now exposed via dedicated endpoints.

Pricing is calculated by looking up the current myst/usd pair value and calculating appropriate values with the provided base config. The base config can be updated via universe(not yet implemented, but soon).

**Note**: Since we're using `latest` tags in node for discovery this will break e2e tests as new flags are required for discovery to run.

### CODE REVIEW

File: price/pricing/pricer.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+	defer p.lock.Unlock()
+
+	return p.updatePrices()
+}
+
+func (p *Pricer) GetPrices() LatestPrices {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if p.lp.CurrentValidUntil.UTC().Before(time.Now().UTC()) {
```
Comment (tomasmik) [inline price/pricing/pricer.go:61]: Make a small helper func `StillValid` just for easier reading?

File: price/api.go [hunk 2]
```diff
@@ -1,56 +1,140 @@
...
+// @Description price config
+// @Product json
+// @Success 200 {array} pricing.Config
+// @Router /prices/config [get]
+// @Tags prices
+func (a *API) GetConfig(c *gin.Context) {
+	cfg, err := a.cfger.Get()
+	if err != nil {
+		log.Err(err).Msg("Failed to get config")
+		c.JSON(http.StatusBadRequest, gorest.NewErrResponse(err.Error()))
```
Comment (tomasmik) [inline price/api.go:52]: If you asked me, I'd say that this is an internal server error 🤷🏻

File: price/api.go [hunk 3]
```diff
@@ -1,56 +1,140 @@
...
+	var cfg pricing.Config
+	if err := c.BindJSON(&cfg); err != nil {
+		c.JSON(http.StatusBadRequest, gorest.NewErrResponse(err.Error()))
+		return
+	}
+
+	err := a.cfger.Update(cfg)
+	if err != nil {
+		log.Err(err).Msg("Failed to update config")
+		c.JSON(http.StatusBadRequest, gorest.NewErrResponse(err.Error()))
```
Comment (tomasmik) [inline price/api.go:76]: This would also be closer to internal server error than bad request.

File: price/api.go [hunk 4]
```diff
@@ -1,56 +1,140 @@
...
 }
+
+func JWTAuthorized(secret string) func(*gin.Context) {
+	return func(c *gin.Context) {
+		authHeader := strings.Split(c.Request.Header.Get("Authorization"), "Bearer ")
+		if len(authHeader) != 2 {
+			c.AbortWithStatusJSON(
+				http.StatusUnauthorized,
+				map[string]string{
+					"error": "Malformed Token",
```
Comment (tomasmik) [inline price/api.go:96]: I guess we aren't breaking the world here, but the error responses will differ depending on if you git an error in this middleware of an actual endpoint. Body here is:
```
map[string]string{"error": "Unauthorized",},
```

Body in the handlers themselves for an error is:
```
Errors []Err `json:"errors"`
```

File: price/pricing/pricer.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+func NewPricer(cfgProvider ConfigProvider, priceAPI FiatPriceAPI, priceLifetime time.Duration, sensibleMystBound Bound) (*Pricer, error) {
+	pricer := &Pricer{
+		cfgProvider:   cfgProvider,
+		priceAPI:      priceAPI,
+		priceLifetime: priceLifetime,
+		mystBound:     sensibleMystBound,
+	}
+	pprotect.CallLoop(func() {
```
Comment (tomasmik) [inline price/pricing/pricer.go:40]: What is the purpose of this call loop with no panic handlers and this not being a new goroutine? Do you expect something to panic that Im missing? If so can `pricer.init` also panic as it's doing the same thing? Or is this just leftover code?

Comment (vkuznecovas) [reply price/pricing/pricer.go:40]: @mdomasevicius ^ :)

File: price/api.go (hunk 3 above)
Comment (vkuznecovas) [reply price/api.go:76]: I'll leave it as a bad request as it returns the validation errors here and I don't feel like complicating the handling here further

File: price/pricing/pricer.go (hunk 5 above)
Comment (mdomasevicius) [reply price/pricing/pricer.go]: sorry left a :hankey:

**Review by tomasmik — APPROVED** [review]

File: cmd/main.go [hunk 6]
```diff
@@ -61,9 +65,31 @@ func main() {
...
+	}
+
+	pricer, err := pricing.NewPricer(
+		cfger,
+		mrkt,
+		time.Minute*5,
+		pricing.Bound{Min: 0.1, Max: 3.0},
+	)
+	if err != nil {
+		log.Err(err).Msg("Failed to initialize Pricer")
```
Comment (tomasmik) [inline cmd/main.go:89]: Just call fatal here

File: price/pricing/pricer.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+func NewPricer(cfgProvider ConfigProvider, priceAPI FiatPriceAPI, priceLifetime time.Duration, sensibleMystBound Bound) (*Pricer, error) {
+	pricer := &Pricer{
+		cfgProvider:   cfgProvider,
+		priceAPI:      priceAPI,
+		priceLifetime: priceLifetime,
+		mystBound:     sensibleMystBound,
+	}
+
+	go schedulePriceUpdate(priceLifetime, pricer)
```
Comment (tomasmik) [inline price/pricing/pricer.go:41]: I would probably go with the good old `Run()` `Stop()` but i guess this works too. 🤷🏻

**Pull request merged**